File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-go/Mailstandard/provider_package_loader.go
package Mailstandard

import "fmt"

type MailProviderPackageLoaderError struct {
	Code    string
	Message string
}

func (e MailProviderPackageLoaderError) Error() string {
	return fmt.Sprintf("%s: %s", e.Code, e.Message)
}

type MailProviderPackageLoadRequest struct {
	ProviderKey     string
	PackageIdentity string
}

type MailResolvedProviderPackageLoadTarget struct {
	PackageEntry MailProviderPackageCatalogEntry
}

type MailProviderModuleNamespace map[string]string
type MailProviderPackageImportFn func(MailResolvedProviderPackageLoadTarget) (MailProviderModuleNamespace, error)
type MailProviderPackageLoader func(MailProviderPackageLoadRequest) (MailProviderModuleNamespace, error)

type MailProviderPackageInstallRequest struct {
	DriverManager any
	LoadRequest   MailProviderPackageLoadRequest
}

func ResolveMailProviderPackageLoadTarget(
	request MailProviderPackageLoadRequest,
) (MailResolvedProviderPackageLoadTarget, error) {
	var packageByProviderKey *MailProviderPackageCatalogEntry
	if request.ProviderKey != "" {
		packageByProviderKey = GetMailProviderPackageByProviderKey(request.ProviderKey)
	}

	var packageByIdentity *MailProviderPackageCatalogEntry
	if request.PackageIdentity != "" {
		packageByIdentity = GetMailProviderPackageByPackageIdentity(request.PackageIdentity)
	}

	if packageByProviderKey != nil && packageByIdentity != nil && packageByProviderKey.PackageIdentity != packageByIdentity.PackageIdentity {
		return MailResolvedProviderPackageLoadTarget{}, MailProviderPackageLoaderError{
			Code:    "provider_package_identity_mismatch",
			Message: "providerKey and packageIdentity must resolve to the same provider package boundary.",
		}
	}

	resolvedPackage := packageByProviderKey
	if resolvedPackage == nil {
		resolvedPackage = packageByIdentity
	}

	if resolvedPackage == nil {
		return MailResolvedProviderPackageLoadTarget{}, MailProviderPackageLoaderError{
			Code:    "provider_package_not_found",
			Message: "No official provider package matches the requested provider boundary.",
		}
	}

	return MailResolvedProviderPackageLoadTarget{PackageEntry: *resolvedPackage}, nil
}

func CreateMailProviderPackageLoader(importPackage MailProviderPackageImportFn) MailProviderPackageLoader {
	return func(request MailProviderPackageLoadRequest) (MailProviderModuleNamespace, error) {
		return LoadMailProviderModule(request, importPackage)
	}
}

func LoadMailProviderModule(
	request MailProviderPackageLoadRequest,
	importPackage MailProviderPackageImportFn,
) (MailProviderModuleNamespace, error) {
	target, err := ResolveMailProviderPackageLoadTarget(request)
	if err != nil {
		return nil, err
	}

	namespace, err := importPackage(target)
	if err != nil {
		return nil, MailProviderPackageLoaderError{
			Code:    "provider_package_load_failed",
			Message: fmt.Sprintf("Reserved provider package loader scaffold could not load %s: %v", target.PackageEntry.PackageIdentity, err),
		}
	}

	if len(namespace) == 0 {
		return nil, MailProviderPackageLoaderError{
			Code:    "provider_module_export_missing",
			Message: "Reserved provider package loader scaffold requires an executable provider module namespace.",
		}
	}

	return namespace, nil
}

func InstallMailProviderPackage(
	request MailProviderPackageInstallRequest,
	importPackage MailProviderPackageImportFn,
) error {
	if _, err := LoadMailProviderModule(request.LoadRequest, importPackage); err != nil {
		return err
	}

	return MailProviderPackageLoaderError{
		Code:    "provider_package_load_failed",
		Message: "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
	}
}

func InstallMailProviderPackages(
	requests []MailProviderPackageInstallRequest,
	importPackage MailProviderPackageImportFn,
) error {
	for _, request := range requests {
		if _, err := LoadMailProviderModule(request.LoadRequest, importPackage); err != nil {
			return err
		}
	}

	if len(requests) > 0 {
		return MailProviderPackageLoaderError{
			Code:    "provider_package_load_failed",
			Message: "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
		}
	}

	return nil
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-rust/src/provider_package_loader.rs
use std::collections::BTreeMap;

use crate::provider_package_catalog::{
    get_mail_provider_package_by_package_identity, get_mail_provider_package_by_provider_key,
    MailProviderPackageCatalogEntry,
};

#[derive(Debug, Clone)]
pub struct MailProviderPackageLoaderError {
    pub code: &'static str,
    pub message: String,
}

impl MailProviderPackageLoaderError {
    pub fn new(code: &'static str, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

#[derive(Debug, Clone, Default)]
#[allow(non_snake_case)]
pub struct MailProviderPackageLoadRequest {
    pub providerKey: Option<String>,
    pub packageIdentity: Option<String>,
}

#[allow(non_snake_case)]
pub struct MailResolvedProviderPackageLoadTarget {
    pub packageEntry: &'static MailProviderPackageCatalogEntry,
}

pub type MailProviderModuleNamespace = BTreeMap<String, String>;
pub type MailProviderPackageImportFn =
    fn(&MailResolvedProviderPackageLoadTarget) -> Result<MailProviderModuleNamespace, MailProviderPackageLoaderError>;
pub type MailProviderPackageLoader = Box<
    dyn Fn(MailProviderPackageLoadRequest) -> Result<MailProviderModuleNamespace, MailProviderPackageLoaderError>,
>;

#[allow(non_snake_case)]
pub struct MailProviderPackageInstallRequest<TDriverManager> {
    pub driverManager: TDriverManager,
    pub loadRequest: MailProviderPackageLoadRequest,
}

pub fn resolve_mail_provider_package_load_target(
    request: &MailProviderPackageLoadRequest,
) -> Result<MailResolvedProviderPackageLoadTarget, MailProviderPackageLoaderError> {
    let package_by_provider_key = request
        .providerKey
        .as_deref()
        .and_then(get_mail_provider_package_by_provider_key);
    let package_by_identity = request
        .packageIdentity
        .as_deref()
        .and_then(get_mail_provider_package_by_package_identity);

    if let (Some(provider_key_entry), Some(package_identity_entry)) =
        (package_by_provider_key, package_by_identity)
    {
        if provider_key_entry.packageIdentity != package_identity_entry.packageIdentity {
            return Err(MailProviderPackageLoaderError::new(
                "provider_package_identity_mismatch",
                "providerKey and packageIdentity must resolve to the same provider package boundary.",
            ));
        }
    }

    let resolved_package = package_by_provider_key.or(package_by_identity).ok_or_else(|| {
        MailProviderPackageLoaderError::new(
            "provider_package_not_found",
            "No official provider package matches the requested provider boundary.",
        )
    })?;

    Ok(MailResolvedProviderPackageLoadTarget {
        packageEntry: resolved_package,
    })
}

pub fn create_mail_provider_package_loader(
    import_package: MailProviderPackageImportFn,
) -> MailProviderPackageLoader {
    Box::new(move |request| load_mail_provider_module(&request, import_package))
}

pub fn load_mail_provider_module(
    request: &MailProviderPackageLoadRequest,
    import_package: MailProviderPackageImportFn,
) -> Result<MailProviderModuleNamespace, MailProviderPackageLoaderError> {
    let target = resolve_mail_provider_package_load_target(request)?;
    let namespace = import_package(&target).map_err(|error| {
        if error.code == "provider_package_load_failed" || error.code == "provider_module_export_missing" {
            error
        } else {
            MailProviderPackageLoaderError::new(
                "provider_package_load_failed",
                format!(
                    "Reserved provider package loader scaffold could not load {}: {}",
                    target.packageEntry.packageIdentity, error.message
                ),
            )
        }
    })?;

    if namespace.is_empty() {
        return Err(MailProviderPackageLoaderError::new(
            "provider_module_export_missing",
            "Reserved provider package loader scaffold requires an executable provider module namespace.",
        ));
    }

    Ok(namespace)
}

pub fn install_mail_provider_package<TDriverManager>(
    request: &MailProviderPackageInstallRequest<TDriverManager>,
    import_package: MailProviderPackageImportFn,
) -> Result<(), MailProviderPackageLoaderError> {
    let _namespace = load_mail_provider_module(&request.loadRequest, import_package)?;

    Err(MailProviderPackageLoaderError::new(
        "provider_package_load_failed",
        "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
    ))
}

pub fn install_mail_provider_packages<TDriverManager>(
    requests: &[MailProviderPackageInstallRequest<TDriverManager>],
    import_package: MailProviderPackageImportFn,
) -> Result<(), MailProviderPackageLoaderError> {
    for request in requests {
        let _namespace = load_mail_provider_module(&request.loadRequest, import_package)?;
    }

    if !requests.is_empty() {
        return Err(MailProviderPackageLoaderError::new(
            "provider_package_load_failed",
            "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
        ));
    }

    Ok(())
}
